enhancements/separating-provider-specific-code-in-the-installer.md

```diff
@@ -0,0 +1,147 @@
+
+### Non-Goals
+
+Remove all provider-specific pieces from the code. That would require a complete refactor of the install config data structure and move the Terraform code. 
+
+## Proposal
+
+### Step 1: Create a Go interface
+
+We should create a Go interface that describes the functionality each provider must implement. For example, from the prototype:
+
+```go
+type PlatformV2 interface {
+	// Name returns the name of the platform.
+	Name() string
+
+	// Metadata translates the installConfig to the appropriate entry in clusterMetadata.
+	Metadata(clusterMetadata *types.ClusterMetadata, installConfig *installconfig.InstallConfig) error
+
+    //...
+}
+```
+
+Not all platforms implement all features. Optional features should be hidden behind a feature flag and a separate interface:
+
+```go
+var ErrNotAnIPIPlatform = errors.New("...")
+
+type PlatformV2 interface {
+    //...
+
+    // GetIPI returns the IPI-specific interface, or ErrNotAnIPIPlatform if the platform does not support IPI.
+	GetIPI() (IPIPlatformV2, error)
+}
+
+type IPIPlatformV2 interface {
+    // IPI-specific 
+}
+``` 
+
+### Step 2: Implement registry
+
+In this step we should implement a registry pattern where providers can register with their names.
+
+### Step 3: Implement feature-switch
+
+Once the registry is complete the installer core code must be changed. Provider-specific switch-case structures can be identified by searching for the usage of the `Name` constants in the `types` package. The existing switch-case should be wrapped in an if-else case. This will help with keeping the functionality of providers that have not yet switched.
+
+```go
+if v2, err := registry.Get(providerName); err == nil {
+    v2.CallSpecifiedFunction(...)
+} else if (!errors.Is(err, registry.ProviderNotFound)) {
+    return err
+} else {
+    switch (...) {
+        //...
+    }
+}
+```
+
+### Step 4: Migrate providers
+
+Providers need to be migrated to the new API.
+
+### Step 5: Remove fallback code
+
+Once all providers are migrated the fallback code can be completely removed.
+
+### Risks and Mitigations
+
+As with every refactor, this change contains the risk of breaking existing functionality. This can be mitigated by migrating only small portions of the code to the new API at a time.
+
+## Design Details
+
+### Open Questions
+
+1. How should the API look in detail?
+2. How should the implementation be merged? All-at-once, or bit by bit?
+
+### Test Plan
+
+Same as testing of the installer for regular releases.
+
+### Graduation Criteria
+
+TBD
+
+### Version Skew Strategy
+
+As the API develops it presents the problem of version skew. As new API calls are introduced they need to either be added to all providers. If that is not possible a new interface should be introduced and a compatibility wrapper should be added. Once all providers are migrated the temporary interface can be removed.
```

Comment:
New function signatures could be implemented as new "optional" interfaces a provider object can implement, with the mechanism of interface assertion.

Reply:
See my comment above, we don't even need to do interface assertion. (Interface assertion is not nice to read and incurs somewhat of a performance penalty.)

Reply:
Why wouldn't it be nice? :grin: Conceptually, it's really what we want here: "dear Platform type, do you implement this method?"

As for the performance penalty: I tend to value readability higher, especially when the total penalty is not expected to pass the 1-second threshold in a 40-minutes process. I would even value aesthetics higher :stuck_out_tongue: